central/declarativeconfig/watch_handler.go

```diff
@@ -1,23 +1,177 @@
+	return &watchHandler{
+		updater:          updater,
+		cachedFileHashes: map[string]md5CheckSum{},
+	}
 }
 
 func (w *watchHandler) OnChange(dir string) (interface{}, error) {
-	return nil, nil
+	entries, err := os.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	declarativeConfigFiles := map[string][]byte{}
+	for _, entry := range entries {
+		if entry.IsDir() {
+			log.Debugf("Found a directory entry within %s: %s. This entry will be skipped", dir, entry.Name())
+			continue
+		}
+		entryContents, err := readDeclarativeConfigFile(path.Join(dir, entry.Name()))
+		if err != nil {
+			log.Errorf("Found an invalid file %s: %+v", entry.Name(), err)
+			continue
+		}
+		declarativeConfigFiles[entry.Name()] = entryContents
+	}
+	return declarativeConfigFiles, nil
 }
 
 func (w *watchHandler) OnStableUpdate(val interface{}, err error) {
+	// We receive an array of file contents (i.e. bytes) which contain valid YAML format (this has been achieved within
+	// OnUpdate, and OnStableUpdate will only be called _iff_ OnStable deemed the contents as valid YAMLs.
+	if err != nil {
+		log.Warnf("Error reading declartive configuration files: %+v", err)
+		return
+	}
+	fileContents, ok := val.(map[string][]byte)
+	if !ok {
+		log.Warnf("Received invalid type in stable update for declarative configuration files: %T", val)
+		return
+	}
+
+	w.logFileContentsNoLock(fileContents)
+
+	if !w.compareHashesForChanges(fileContents) && !w.checkForDeletedFiles(fileContents) {
+		log.Debugf("Found no changes from before in content, no reconciliation will be triggered")
+		return
+	}
+
+	log.Debugf("Found changes in declarative configuration files, reconciliation will be triggered")
+	w.mutex.RLock()
+	defer w.mutex.RUnlock()
+	w.updater.ReconcileDeclarativeConfigs(maputil.Values(fileContents))
 }
 
 func (w *watchHandler) OnWatchError(err error) {
 	log.Errorf("Error watching declarative configuration directory: %+v", err)
 }
+
+// compareHashesForChanges compares the file contents for changes based on previous hashes stored for this handler.
+// Additionally, it will also verify if the cache contains additional files that are not included within the given
+// file contents. In the end, the cached values will be changed to reflect the passed file contents.
+// It will return true if:
+//   - any of the file hashes changed from the cached value to the value in the given file contents.
+//   - a file which previously was not part of the cached values is contained within the file contents.
+//
+// Otherwise, it will return false.
+func (w *watchHandler) compareHashesForChanges(fileContents map[string][]byte) bool {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	var changedFiles bool
+	for fileName, fileContent := range fileContents {
+		cachedHash, ok := w.cachedFileHashes[fileName]
+		fileHash := md5.Sum(fileContent)
+		if !ok || !bytes.Equal(cachedHash[:], fileHash[:]) {
+			w.cachedFileHashes[fileName] = fileHash
```

Comment:
My understanding is that this code works under the assumption that the file is always reconciled error-free after this line. Currently, that's the case(there are no errors returned all until line 90:
https://github.com/stackrox/stackrox/pull/4828/files#diff-e828eface4a18e3746716e6b6ff4135d5773e64d644cdd10eff0ca28664e1394R90

Perhaps, it would be easier not to miss any changes if we update the hash value after it actual reconciliation?

Reply:
So, essentially you would want to have an error returned from `ReconcileDeclarativeConfigs` and only _then_ update the cached file hashes?

Not sure I'd follow the same, since these watch handlers are for me just purely for reading the file content as YAML bytes and passing them to the declarative configuration manager for reconciliation.

What exactly would make the difference in doing it for a _successfuly_ reconciliation, which could take longer and might be asynchronous (right now, we don't actually do anything in `ReconcileDeclarativeConfigs`, but I doubt its going to be synchronous going forward, as this essentially would add the transformed protos to some sort of loop for reconciliation).

Reply:
Let's keep it as it is. It's important though to add a comment that there should be no error-prone code between these lines
https://github.com/stackrox/stackrox/pull/4828/files#diff-e828eface4a18e3746716e6b6ff4135d5773e64d644cdd10eff0ca28664e1394R80-R88 -- otherwise, it's possible changes to files will not be picked up.